File: pi-socket/src/log.ts
/**
 * Structured logger for pi-socket.
 *
 * pi-socket runs inside pi's Node.js process. console.log/error go to
 * pi's TUI and flood the user. stderr is also pi's TUI. So we log to
 * a JSONL file at ~/.pi/logs/pi-socket.jsonl.
 *
 * Log levels:
 * - info:  Normal operations — startup, connections, registrations.
 * - warn:  Expected degraded conditions — reconnecting, client dropped.
 * - error: Unanticipated errors caught by the safety net. Each error
 *          entry has needsHardening: true and represents a gap in the
 *          inner error-handling layer that should be fixed in code.
 *
 * The harden-pi-socket skill reads entries where needsHardening is true,
 * cross-references with its ledger, and proposes targeted code fixes.
 */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

const LOG_DIR = path.join(os.homedir(), ".pi", "logs");
const LOG_FILE = path.join(LOG_DIR, "pi-socket.jsonl");

/** Exported so the hardening skill knows where to look. */
export const LOG_PATH = LOG_FILE;

export type LogLevel = "info" | "warn" | "error";

interface LogEntry {
  ts: string;
  level: LogLevel;
  component: string;
  msg: string;
  needsHardening?: true;
  boundary?: string;
  error?: string;
  stack?: string;
  [key: string]: unknown;
}

let initialized = false;

function ensureDir(): void {
  if (initialized) return;
  try {
    fs.mkdirSync(LOG_DIR, { recursive: true });
    initialized = true;
  } catch {
    // Can't create log dir — logging is best-effort.
  }
}

function write(entry: LogEntry): void {
  ensureDir();
  try {
    fs.appendFileSync(LOG_FILE, JSON.stringify(entry) + "\n");
  } catch {
    // Best-effort — do NOT throw.
  }
}

/**
 * Log a normal operational event.
 */
export function info(component: string, msg: string, data?: Record<string, unknown>): void {
  write({ ts: new Date().toISOString(), level: "info", component, msg, ...data });
}

/**
 * Log an expected degraded condition (not a bug, but worth noting).
 */
export function warn(component: string, msg: string, data?: Record<string, unknown>): void {
  write({ ts: new Date().toISOString(), level: "warn", component, msg, ...data });
}

/**
 * Log an unanticipated error caught by the safety net.
 * Marked with needsHardening: true for the hardening skill to process.
 */
export function error(boundary: string, err: unknown, data?: Record<string, unknown>): void {
  write({
    ts: new Date().toISOString(),
    level: "error",
    component: "pi-socket",
    msg: err instanceof Error ? err.message : String(err),
    needsHardening: true,
    boundary,
    error: err instanceof Error ? err.message : String(err),
    stack: err instanceof Error ? err.stack : undefined,
    ...data,
  });
}

File: hypivisor/src/log.rs
//! Structured JSONL logger for hypivisor.
//!
//! Writes to `~/.pi/logs/hyper-pi.jsonl` — the same unified log that
//! pi-socket uses. Each entry includes a `component` field to distinguish
//! the source. Entries with `level: "error"` are marked `needsHardening: true`
//! so the harden skill can process them.
//!
//! This is **in addition to** the `tracing` macros that go to stderr.
//! stderr gives real-time visibility; the JSONL file gives persistent
//! structured logging for the harden skill.

use chrono::Utc;
use serde_json::json;
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::sync::OnceLock;

fn log_dir() -> &'static PathBuf {
    static DIR: OnceLock<PathBuf> = OnceLock::new();
    DIR.get_or_init(|| {
        let dir = dirs::home_dir()
            .unwrap_or_else(|| ".".into())
            .join(".pi")
            .join("logs");
        let _ = fs::create_dir_all(&dir);
        dir
    })
}

fn write_entry(entry: serde_json::Value) {
    let path = log_dir().join("hyper-pi.jsonl");
    if let Ok(mut file) = OpenOptions::new().create(true).append(true).open(&path) {
        // Each line is well under PIPE_BUF (4096), so append is atomic
        let _ = writeln!(file, "{}", entry);
    }
}

/// Log a normal operational event.
pub fn info(component: &str, msg: &str) {
    write_entry(json!({
        "ts": Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        "level": "info",
        "component": component,
        "msg": msg,
    }));
}

/// Log an expected degraded condition (not a bug, but worth noting).
pub fn warn(component: &str, msg: &str) {
    write_entry(json!({
        "ts": Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        "level": "warn",
        "component": component,
        "msg": msg,
    }));
}

/// Log an unanticipated error. Marked `needsHardening: true` for the harden skill.
pub fn error(boundary: &str, msg: &str) {
    write_entry(json!({
        "ts": Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        "level": "error",
        "component": "hypivisor",
        "msg": msg,
        "needsHardening": true,
        "boundary": boundary,
    }));
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn write_entry_creates_valid_jsonl() {
        let dir = std::env::temp_dir().join("hypi_log_test");
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        let path = dir.join("test.jsonl");

        let entry = json!({
            "ts": "2026-01-01T00:00:00.000Z",
            "level": "info",
            "component": "test",
            "msg": "hello",
        });

        if let Ok(mut file) = OpenOptions::new().create(true).append(true).open(&path) {
            let _ = writeln!(file, "{}", entry);
        }

        let content = fs::read_to_string(&path).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(content.trim()).unwrap();
        assert_eq!(parsed["level"], "info");
        assert_eq!(parsed["msg"], "hello");

        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn error_entry_has_needs_hardening() {
        let entry = json!({
            "ts": "2026-01-01T00:00:00.000Z",
            "level": "error",
            "component": "hypivisor",
            "msg": "test error",
            "needsHardening": true,
            "boundary": "test.boundary",
        });

        assert_eq!(entry["needsHardening"], true);
        assert_eq!(entry["level"], "error");
        assert_eq!(entry["boundary"], "test.boundary");
    }
}
